### DISCUSSION TITLE
Html Entities (&rarr;)

### DISCUSSION THREAD

**Discussion by robacarp** [stephannv/blueprint#81] [Q&A] [answered]
The crystal [HTML.escape](https://crystal-lang.org/api/1.14.0/HTML.html#escape%28string%3AString%29%3AString-class-method) which [blueprint uses](https://github.com/stephannv/blueprint/blob/main/src/blueprint/html/buffer_appender.cr#L3) isn't as smart as it could be. It'll [stubbornly re-escape things which are already escaped](https://play.crystal-lang.org/#/r/hajp):

```
"Before: &rarr; and After: #{HTML.escape("&rarr;")}"
#=> "Before: &rarr; and After: &amp;rarr;"
```

Rails [provides something clever](https://github.com/rails/rails/blob/a11f0a63673d274c59c69c2688c63ba303b86193/actionview/lib/action_view/helpers/tag_helper.rb#L576-L585) which will only escape things which don't appear to already be escaped:

```
escape_once("1 < 2 &amp; 3")
# => "1 &lt; 2 &amp; 3"

escape_once("&lt;&lt; Accept & Checkout")
# => "&lt;&lt; Accept &amp; Checkout"
```

This functionality exists on top of the ruby escape functionality, which is almost certainly what Crystal's is based off of.

For now, I'm able to render an html entity manually in blueprint with `raw safe "&larr;"` -- but that'll get significantly more complicated if I have markdown content with html entities encoded in it already.

**Comment by stephannv** [comment, marked as answer]
I will investigate this issue. I tried to replicate the Rails behavior:
```crystal
HTML_ESCAPE = {
  "&" => "&amp;",
  ">" => "&gt;",
  "<" => "&lt;",
  %(") => "&quot;",
  "'" => "&#39;"
}
HTML_ESCAPE_ONCE_REGEXP = /["><']|&(?!([a-zA-Z]+|(#\d+)|(#[xX][\dA-Fa-f]+));)/

value.gsub(HTML_ESCAPE_ONCE_REGEXP, HTML_ESCAPE)
```

But Blueprint got much more slower than using HTML.escape (on benchmark it went from 8x slower than ECR to 14x slower). I will need tweak some things, eg. use faster escaping for attribute values and use this escape_once behavior for content.

**Reply by stephannv** [reply]
@robacarp  what do you think about https://github.com/stephannv/blueprint/pull/83 ?

I kept the default Crystal escaper and added a new one that you can use through the `escape_once` helper.

Usage:
```crystal
def blueprint
  plain escape_once("1 < 2 &amp; 3")
  
  span { escape_once("&lt;&lt; Accept & Checkout") }
  
  div escape_once("1 < 2 &amp; 3")
end
```
Results:
```html
1 &lt; 2 &amp; 3

<span>&lt;&lt; Accept &amp; Checkout</span>

<div>1 &lt; 2 &amp; 3</div>
```

**Reply by robacarp** [reply]
I'm in favor -- it seems like it plays nicely with the existing safety framework too, which is great. I wonder if this should be a crystal stdlib feature?

**Reply by stephannv** [reply]
Thanks for the feedback. `escape_once` is live on [0.11.0](https://github.com/stephannv/blueprint/releases/tag/v0.11.0)

I'm not sure about how the core team decide what should be part of stdlib or not, but since it is just 3 lines, it is easy to implement it by ourselves. But if other languages have this feat built in, it should easier to convince core team members to accept the proposal.

**Reply by robacarp** [reply]
amazing, thank you!
